functional/src/Logic.hs

```diff
@@ -12,29 +13,26 @@ switchPlayer game =
       PlayerX -> game { gamePlayer = PlayerO }
       PlayerO -> game { gamePlayer = PlayerX }
 
-playerWon :: Player -> Board -> Bool
-playerWon player board = any isVictoryProj projs
-    where projs = allRowCoords
-                  ++ allColumnCoords
-                  ++ allDiagCoords
-          allRowCoords = [[(i, j) | j <- [0 .. n - 1]] | i <- [0 .. n - 1]]
-          allColumnCoords = [[(j, i) | j <- [0 .. n - 1]] | i <- [0 .. n - 1]]
-          allDiagCoords = [ [(i, i) | i <- [0 .. n - 1]]
-                          , [(i, n - 1 - i) | i <- [0 .. n - 1]]
-                          ]
-          isVictoryProj proj = (n ==)
-                               $ length
-                               $ filter (\cell -> cell == Full player)
-                               $ map (\coord -> board ! coord) proj
+full :: [Cell] -> Maybe Player
+full (x@(Just player):xs) | and $ map (== x) xs = Just player
+full _                                          = Nothing
+
+winner :: Board -> Maybe Player
+winner board = asum $ map full cells
+    where cells = map (map (board !)) coords :: [[Cell]]
+          coords = rows ++ cols ++ diags
+
+          rows  = [[(i,j) | i <- [0..n-1]] | j <- [0..n-1]]
```

Comment:
Can we for each of the "projections" down there write something like `[[board ! (i,j) | i <- [0..n-1]] | j <- [0..n-1]]`? And the `cells` above is going to be simply the concatenation of all of the projections: `cells = rows ++ cols ++ diags`. That may help to get rid of that awkward nested `map` above. What do you think?